### PR TITLE
Add saturate filter

### LINKED ISSUE #221: Feature request: Saturation
I would like to ask for image colour saturation. When paired with blur it gives a really good effect for modern image placeholders.

I can even chime in and add a pull request for this feature if you don't have the time to implement it.

### PR DESCRIPTION
**Pull request by Tylerian**
This pull request adds support for image color saturation.

### CODE REVIEW

**Comment by Tylerian** [comment]
I finished working on this pull request, awaiting feedback :)

**Comment by Tylerian** [comment]
@codecov-io

**Comment by Tylerian** [comment]
@kleisauke It's been almost a month since I've implemented this feature, any plans on merging this?

File: src/api/api_manager_impl.cpp [hunk 1]
```diff
@@ -154,14 +155,14 @@ utils::Status ApiManagerImpl::process(const std::string &query,
 
     // Image processing phase 3 (adjustments, effects, etc.)
     image = image | embed | rotation | brightness | contrast | gamma | sharpen |
-            filter | blur | tint | background | mask;
+            filter | blur | tint | background | mask | saturate;
```
Comment (kleisauke) [inline src/api/api_manager_impl.cpp:158]: Saturation should not be done at the last stage. It should probably be done between the `brightness` and `tint` operation.

File: src/api/api_manager_impl.cpp [hunk 2]
```diff
@@ -154,14 +155,14 @@ utils::Status ApiManagerImpl::process(const std::string &query,
...
 
     // Write the image to a target
     stream.write_to_target(image, target);
 
     // Clean up libvips' per-request data and threads
     clean_up();
-
+
```
Comment (kleisauke) [inline src/api/api_manager_impl.cpp:165]: nitpick: unnecessary whitespace

File: src/api/processors/saturate.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+                .colourspace(VIPS_INTERPRETATION_sRGB)
+                .recomb(image.bands() == 3
+                        ? VImage::new_from_memory(saturate.begin(), 9 * sizeof(double), 3,
+                                                  3, 1, VIPS_FORMAT_DOUBLE)
+                        : VImage::new_matrixv(4, 4,
+                                              saturate[0], saturate[1], saturate[2], 0.0,
+                                              saturate[3], saturate[4], saturate[5], 0.0,
+                                              saturate[6], saturate[7], saturate[8], 0.0,
+                                              0.0, 0.0, 0.0, 1.0));
+    // clang-format on
```
Comment (kleisauke) [inline src/api/processors/saturate.cpp:38]: How about exposing libvips' recombination matrix operation instead? Something like this:
```suggestion
    auto recomb = query_->get<std::vector<float>>("recomb", {});

    // Should we process the image?
    if (recomb.empty()) {
        return image;
    }

    // Need to convert std::vector<float> to std::vector<double>
    std::vector<double> m(recomb.begin(), recomb.end());

    // Recomb the image with the specified matrix.
    // clang-format off
    auto matrix = image.bands() == 3
                      ? VImage::new_from_memory(m.data(), 9 * sizeof(double), 3,
                                                3, 1, VIPS_FORMAT_DOUBLE)
                      : VImage::new_matrixv(4, 4,
                                            m[0], m[1], m[2], 0.0,
                                            m[3], m[4], m[5], 0.0,
                                            m[6], m[7], m[8], 0.0,
                                            0.0, 0.0, 0.0, 1.0);

    return image
        .colourspace(VIPS_INTERPRETATION_sRGB)
        .recomb(matrix);
    // clang-format on
```

So that users can fill in a 3x3 recombination matrix of their choice. For example, a sepia filter could look like this:
```
https://images.weserv.nl/static/zebra.jpg?recomb=0.3588,0.7044,0.1368,0.2990,0.5870,0.1140,0.2392,0.4696,0.0912
```

This is effectively the same as `<feColorMatrix>` in SVG.

File: test/api/fixtures.h [hunk 4]
```diff
@@ -51,6 +51,9 @@ struct Fixtures {
 
     std::string input_jpg_overlay_layer_2{dir + "/alpha-layer-2-ink.jpg"};
 
+    // https://images.unsplash.com/photo-1586015910630-34cc709e5f91
+    std::string input_jpg_saturate{dir + "/photo-1586015910630-34cc709e5f91.png"}
+
```
Comment (kleisauke) [inline test/api/fixtures.h:56]: Could you not use [`input_jpg`](https://raw.githubusercontent.com/weserv/images/5.x/test/api/fixtures/2569067123_aca715a2ee_o.jpg) instead? I don't see the point of including another test image.

File: test/api/fixtures.h [hunk 5]
```diff
@@ -81,6 +84,9 @@ struct Fixtures {
     // https://www.flickr.com/photos/grizdave/2569067123/ (same as input_jpg)
     std::string input_png_rgb_with_alpha{dir + "/2569067123_aca715a2ee_o.png"};
 
+    // https://images.unsplash.com/photo-1586015910630-34cc709e5f91 (same as input_jpg_saturate)
+    std::string input_png_saturate{dir + "/photo-1586015910630-34cc709e5f91.png"}
+
```
Comment (kleisauke) [inline test/api/fixtures.h:89]: Could you not use [`input_png_rgb_with_alpha`](https://raw.githubusercontent.com/weserv/images/5.x/test/api/fixtures/2569067123_aca715a2ee_o.png) instead?

File: src/api/parsers/query.cpp [hunk 6]
```diff
@@ -60,6 +60,7 @@ const TypeMap &type_map = {
         {"loop",    typeid(int)},               // TODO(kleisauke): Documentation needed.
         {"delay",   typeid(std::vector<int>)},  // TODO(kleisauke): Documentation needed.
         {"fsol",    typeid(bool)},              // TODO(kleisauke): Documentation needed.
+        {"sat",     typeid(float)},
```
Comment (kleisauke) [inline src/api/parsers/query.cpp:63]: If we're going to turn this into a recombination operation, this needs to be changed as well:
```suggestion
        {"recomb",  typeid(std::vector<float>)},
```

And something like this needs to be included within `add_value`:
```cpp
    } else if (key == "recomb") {  // type == typeid(std::vector<float>)
        auto matrix = tokenize<float>(value, ",", 9);

        if (matrix.size() == 9) {
            map.emplace("recomb", matrix);
        }
```

**Review by kleisauke — CHANGES_REQUESTED** [review]
Thanks for the PR Jairo and my apologies for the delay in reviewing this.

I've created a separate PR for the LCH colourspace approach in #226, as I think this is more correct. StackOverflow also has a good answer on this matter: https://stackoverflow.com/a/19325417/10952119
(That's for CSS `hue-rotate()` but also applies to `saturate()`)

If users want to match the behavior of CSS, how about changing this PR to expose [libvips' recombination matrix operation](https://libvips.github.io/libvips/API/current/libvips-conversion.html#vips-recomb) instead (which is effectively the same as `<feColorMatrix>` in SVG)? Users are then able to fill in a 3x3 recombination matrix of their choice. For example, a sepia filter (`&filt=sepia`) could look like this:
```
https://images.weserv.nl/static/zebra.jpg?recomb=0.3588,0.7044,0.1368,0.2990,0.5870,0.1140,0.2392,0.4696,0.0912
```

If you want, I could make a separate PR for this as well.

**Comment by kleisauke** [comment]
The deletion of the `true-streaming` branch triggered the automatic closing of this PR. My apologies. This was an unexpected side effect.

If you're still interested in exposing [libvips' recombination matrix operation](https://libvips.github.io/libvips/API/current/libvips-conversion.html#vips-recomb), please open a new PR, as GitHub won't let me change the destination branch or reopen this PR.
